jetstream_consumer.go

```diff
@@ -0,0 +1,611 @@
+		}
+		o.MaxDeliver = n
+		return nil
+	}
+}
+
+// FilterStreamBySubject filters the messages in a wildcard stream to those matching a specific subject
+func FilterStreamBySubject(s string) ConsumerConfigOption {
+	return func(o *ConsumerConfig) error {
+		o.FilterSubject = s
+		return nil
+	}
+}
+
+// ReplayInstantly delivers messages to the consumer as fast as possible
+func ReplayInstantly() ConsumerConfigOption {
+	return func(o *ConsumerConfig) error {
+		o.ReplayPolicy = ReplayInstant
+		return nil
+	}
+}
+
+// ReplayAsReceived delivers messages at the rate they were received at
+func ReplayAsReceived() ConsumerConfigOption {
+	return func(o *ConsumerConfig) error {
+		o.ReplayPolicy = ReplayOriginal
+		return nil
+	}
+}
+
+// SamplePercent configures sampling of a subset of messages expressed as a percentage
+func SamplePercent(i int) ConsumerConfigOption {
+	return func(o *ConsumerConfig) error {
+		if i < 0 || i > 100 {
+			return fmt.Errorf("sample percent must be 0-100")
+		}
+
+		if i == 0 {
+			o.SampleFrequency = ""
+			return nil
+		}
+
+		o.SampleFrequency = fmt.Sprintf("%d%%", i)
+		return nil
+	}
+}
+
+// RateLimitBitsPerSecond limits message delivery to a rate in bits per second
+func RateLimitBitsPerSecond(bps uint64) ConsumerConfigOption {
+	return func(o *ConsumerConfig) error {
+		o.RateLimit = bps
+		return nil
+	}
+}
+
+// NewConsumerConfig creates a new consumer configuration based on cfg modified by opts
+func NewConsumerConfig(cfg ConsumerConfig, opts ...ConsumerConfigOption) (ConsumerConfig, error) {
+	res := ConsumerConfig{}
+
+	for _, opt := range opts {
+		err := opt(&res)
+		if err != nil {
+			return ConsumerConfig{}, err
+		}
+	}
+
+	return res, nil
+}
+
+func (nc *Conn) createOrUpdateConsumer(opts *jsOpts) (*ConsumerInfo, error) {
+	if opts.streamName == "" {
+		// TODO make err constants
+		return nil, fmt.Errorf("nats: Stream name is required")
+	}
+	if opts.consumer == nil {
+		// TODO make err constants
+		return nil, fmt.Errorf("nats: Consumer configuration is required")
+	}
+
+	crj, err := json.Marshal(&JSApiConsumerCreateRequest{
+		Stream: opts.streamName,
+		Config: *opts.consumer,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := opts.context(nc.Opts.JetStreamTimeout)
+	defer cancel()
+
+	var subj string
+	switch len(opts.consumer.Durable) {
+	case 0:
+		subj = fmt.Sprintf(JSApiConsumerCreateT, opts.streamName)
+	default:
+		subj = fmt.Sprintf(JSApiDurableCreateT, opts.streamName, opts.consumer.Durable)
+	}
+
+	resp, err := nc.RequestWithContext(ctx, subj, crj)
+	if err != nil {
+		return nil, err
+	}
+
+	cresp := &JSApiConsumerCreateResponse{}
+	err = json.Unmarshal(resp.Data, cresp)
+	if err != nil {
+		return nil, err
+	}
+
+	if cresp.IsError() {
+		return nil, cresp.ToError()
+	}
+
+	return cresp.ConsumerInfo, nil
+}
+
+const (
+	JSApiConsumerCreateT                   = "$JS.API.CONSUMER.CREATE.%s"
+	JSApiDurableCreateT                    = "$JS.API.CONSUMER.DURABLE.CREATE.%s.%s"
```

Comment:
This feels like management lib/module which should be separate no?

Reply:
Yes did consider trimming them but having these here will help people who want to use Request() to interact with the API for things we don’t have here. So kept them but also happy to say people should call into jsm.go for that